buildtools/maven/maven.go

```diff
@@ -59,56 +137,153 @@ func (m *Maven) Compile(dir string) error {
+	}
+
+	if checked[absPath] {
+		return nil, nil
+	}
+
+	checked[pomFile] = true
+
+	var pom Manifest
+	if err := files.ReadXML(&pom, pomFile); err != nil {
+		return nil, errors.Wrapf(err, "could not read POM file %q", pomFile)
+	}
+
+	modules := make([]MvnModule, 1, 1+len(pom.Modules))
+
+	name := pom.Name
+	if name == "" {
+		name = pom.ArtifactID
 	}
 
-	var modules []string
+	modules[0] = MvnModule{Name: name, Target: pomFile, Dir: dir}
+
 	for _, module := range pom.Modules {
-		children, err := Modules(filepath.Join(dir, module))
+		childPath := filepath.Join(dir, module)
+		children, err := Modules(childPath, baseDir, checked)
 		if err != nil {
-			return nil, errors.Wrap(err, "could not read child modules")
+			return nil, errors.Wrapf(err, "could not read child module at %q", childPath)
 		}
 		modules = append(modules, children...)
 	}
 
-	groupID := pom.GroupID
-	if groupID == "" {
-		groupID = pom.Parent.GroupID
-	}
-	modules = append(modules, groupID+":"+pom.ArtifactID)
-
 	return modules, nil
 }
 
-func (m *Maven) DependencyList(dir string) (string, error) {
-	output, _, err := exec.Run(exec.Cmd{
-		Name: m.Cmd,
-		Dir:  dir,
-		Argv: []string{"dependency:list", "--batch-mode"},
-	})
-	return output, err
+// DependencyList runs Maven's dependency:list goal for the specified project.
+func (m *Maven) DependencyList(dir, buildTarget string) (string, error) {
+	return m.tryDependencyCommands("list", dir, buildTarget)
 }
 
-func (m *Maven) DependencyTree(dir, project string) ([]Dependency, map[Dependency][]Dependency, error) {
-	output, _, err := exec.Run(exec.Cmd{
-		Name: m.Cmd,
-		Dir:  dir,
-		Argv: []string{"dependency:tree", "--batch-mode", "--projects", project},
-	})
+// DependencyTree runs Maven's dependency:tree goal for the specified project.
+func (m *Maven) DependencyTree(dir, buildTarget string) (graph.Deps, error) {
+	output, err := m.tryDependencyCommands("tree", dir, buildTarget)
 	if err != nil {
-		return nil, nil, err
+		return graph.Deps{}, err
 	}
-
 	return ParseDependencyTree(output)
 }
 
+func (m *Maven) tryDependencyCommands(subGoal, dir, buildTarget string) (stdout string, err error) {
+	goalArgs := []string{"dependency:" + subGoal, "--batch-mode"}
+	cmd := exec.Cmd{Name: m.Cmd, Dir: dir}
+
+	// First, we try with the buildTarget because we have no idea what it is or whether the reactor already
+	// knows about the module.
+	cmd.Argv = append(goalArgs, "--projects", buildTarget)
+	output, _, err := exec.Run(cmd)
+	if err != nil {
+		// Now we try to identify the groupId:artifactId identifier for the module and specify the path to
+		// the manifest file directly.
+		pom, err2 := m.ResolveManifestFromBuildTarget(buildTarget)
+		if err2 != nil {
+			// Using buildTarget as a module ID or as a path to a manifest did not work.
+			// Return just the error from running the mvn goal the first time.
+			return "", errors.Wrap(err, "could not use Maven to list dependencies")
```

Comment:
Should this fallback instead of return an error?

Reply:
there's nothing really to fall back to, since this is already the fallback method: when Maven cannot take our original `buildTarget` we try again but with `--project=groupId:artifactId` and `--file=the-pom-file.xml`. If we can't identify the groupId or artifactId from the POM file, then there's nothing else we can do.